proposals/2019-08-26-json-module.md

```diff
@@ -0,0 +1,99 @@
+updated: 2019-08-26
+authors:
+  - [jmillikin](https://john-millikin.com)
+reviewers:
+  - Starlark core reviewers
+discussion thread: [#83](https://github.com/bazelbuild/starlark/pull/83)
+---
+
+# JSON Module
+
+## Abstract
+
+This document proposes an API for encoding and decoding JSON from Starlark. If accepted, implementations of Starlark
+that implement JSON support would be expected to implement this API. The goal is to allow users to write JSON code that
+behaves consistently across Starlark implementations.
+
+## Background
+
+JSON is a popular syntax for representing basic data structures. Partial support for parsing or serializing JSON from
+Starlark has been independently requested and/or implemented several times:
+
+* Bazel's [`struct`](https://docs.bazel.build/versions/master/skylark/lib/struct.html) type has a `to_json()` method
+  that can generate JSON documents. Issue [bazelbuild/bazel#7879](https://github.com/bazelbuild/bazel/issues/7879)
+  requests additional control over the behavior of this method (i.e. optional whitespace).
+
+* [bazelbuild/bazel#3732](https://github.com/bazelbuild/bazel/issues/3732) requests support for parsing JSON from
+  within a Bazel repository rule, for use with introspecting the state of external tools.
+
+* [`json_parser.bzl`](https://github.com/erickj/bazel_json) is a JSON parser implemented entirely within Starlark.
+  Limitations of the Starlark language prevent this parser from being recommended for production use.
+
+* [google/starlark-go#179](https://github.com/google/starlark-go/pull/179) proposes a JSON module for the Go
+  implementation of starlark.
+
+* [Skycfg](https://github.com/stripe/skycfg) has a `json.marshal()` method that can generate JSON documents, for use
+  with config formats based on JSON syntax.
+
+JSON implementations in the broader community have wildly variant APIs. It is likely that ad-hoc JSON extensions to
+Starlark will have different APIs between Starlark implementations.
+
+## Proposed API
+
+The JSON module is a value named `json` in the global namespace. Its API comprises the following functions. Starlark
+implementations are not required to support the entire API, but should avoid extending the API with non-standard
+functions or parameters.
+
+JSON implementations should comply with the format documented as
+[ECMA-404](https://www.ecma-international.org/publications/standards/Ecma-404.htm) and
+[RFC 8259](https://www.rfc-editor.org/rfc/rfc8259.html), which supercedes earlier drafts of JSON. In particular, the
+only permitted character encoding for contemporary JSON is UTF-8.
+
+To maintain compatibility with existing callers, new required parameters should not be added to these functions. New
+optional parameters should be defined using keyword-only parameters
+([PEP-3102](https://www.python.org/dev/peps/pep-3102/)).
+
+### json.decode()
+
+The `json.decode()` function decodes a JSON document into a Starlark value.
+
+```python
+def json.decode(data):
```

Comment:
Very cool! Is there any reason we wouldn't follow the [python json interface](https://docs.python.org/3/library/json.html), perhaps only the string `loads`/`dumps` versions? They might have a few extra flags, but 99% of the uses would suffice with the flags already included in this spec.

Reply:
I'm fine with following the Python API here. Historically Bazel hasn't worried too much about cross-compatibility with the Python stdlib, and I thought `loads` / `dumps` is a little opaque to people who don't use Python regularly, but I'd accept the functionality under ~any name.